Pbh cli

vadymhlushko-mlnx opened the pull request:
<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did
The Policy-Based Hashing CLI

#### How I did it

#### How to verify it

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

stepanblyschak commented on scripts/fast-reboot:
```diff
@@ -483,7 +512,9 @@ systemctl stop swss
 if [[ "$REBOOT_TYPE" = "warm-reboot" || "$REBOOT_TYPE" = "fastfast-reboot" ]]; then
     initialize_pre_shutdown
 
+    check_issu_bank_file
```
if [[ "x$sonic_asic_type" == x"mellanox" ]]

vadymhlushko-mlnx replied:
Fixed

stepanblyschak commented on scripts/fast-reboot:
```diff
@@ -483,7 +512,9 @@ systemctl stop swss
 
+    check_issu_bank_file
     request_pre_shutdown
+    check_issu_bank_file
```
if [[ "x$sonic_asic_type" == x"mellanox" ]]

vadymhlushko-mlnx replied:
Fixed

stepanblyschak commented on scripts/fast-reboot:
```diff
@@ -156,6 +156,35 @@ function request_pre_shutdown()
+function abort_script()
+{
+    error "Abort warm-reboot ..."
+    exit "${EXIT_FAILURE}"
```
Why we need another function?

stepanblyschak commented on scripts/fast-reboot:
```diff
@@ -156,6 +156,35 @@ function request_pre_shutdown()
+function abort_script()
+{
+    error "Abort warm-reboot ..."
+    exit "${EXIT_FAILURE}"
```
Use dedicated return code, e.g. mlnx FW burn error uses  MLNX_EXIT_FW_ERROR

vadymhlushko-mlnx replied:
Fixed

stepanblyschak commented on scripts/fast-reboot:
```diff
@@ -156,6 +156,29 @@ function request_pre_shutdown()
+    issu_file_content=`awk '{print $0}' ${ISSU_BANK_FILE}`
+    if [[ "$issu_file_content" != "0" && "$issu_file_content" != "1" ]]; then
+        error "(${ISSU_BANK_FILE}) has invalid content ..."
+        exit "${MLNX_ISSU_BANK_BROKEN}"
```
MLNX_ISSU_BANK_BROKEN is it resolved in this context? It is defined in another function

vadymhlushko-mlnx replied:
Fixed

stepanblyschak commented:
I think that if issu_bank.txt is broken, we can restore it manually, but needs to be verified by warm-reboot test.

```
admin@arc-switch1004:~$ docker exec -it syncd sx_api_dbg_generate_dump.py
[+] opening sdk
May 13 16:12:09 INFO    LOG: Initializing SX log with STDOUT as output file.
sx_api_open handle:0x559fabe202f4 , rc 0
No dump file path was entered as a parameter.
Will use the default: /tmp/sdkdump path.
sx_api_dbg_generate_dump, rc 0
admin@arc-switch1004:~$ docker exec -it syncd cat /tmp/sdkdump | grep 'ISSU bank'
admin@arc-switch1004:~$ docker exec -it syncd cat /tmp/sdkdump | grep 'ISSU Bank'
ISSU Bank                                0
admin@arc-switch1004:~$ printf 0 > /host/warmboot/issu_bank.txt
```

aravindmani-1 committed:
Stop PMON docker before cold and soft reboots (#1514)

Prevent potential kernel oops if drivers are removed/devices are deinitialized while PMon daemons are still trying to access those devices.

smaheshm committed:
[techsupport] Update show ip interface command (#1562)

+ update 'show ip interface' command in tech support collection
+ fix unbound variable issue in case command times out

renukamanavalan committed:
1) Loopback interfaces with valid nexthop IP are not ignored/treated as loopback. (#1565)

2) The vrf routes are *not* handled.

smaheshm committed:
[CLI][queue counters] add JSON output option for queue counters (#1505)

+ added tests for 'show queue counters' CLI

renukamanavalan committed:
No more IP validation as it is more likely a URL (#1555)

Dropped IP validation as server takes URL.

daall committed:
[acl] Use a list instead of a comma-separated string for ACL port list (#1519)

Signed-off-by: Danny Allen <daall@microsoft.com>

santhosh-kt committed:
platform pre-check for reboot in master branch (#1556)

What I did
Added platform pre check support in reboot script.
Checking platform based changes before stopping dockers and sonic services.
Porting changes in master from 201911 branch #1472
How I did it
On branch reboot_pre_check_master
Changes not staged for commit:
(use "git add ..." to update what will be committed)
(use "git checkout -- ..." to discard changes in working directory)

modified:   scripts/reboot
How to verify it
Write a platform pre check script(platform_reboot_pre_check) and place it in /usr/share/sonic/device// directory.
If the script exit with status 0, reboot will be proceeded.
If script exit with non-zero status, the reboot script gets stopped.

a-barboza committed:
RADIUS Management User Authentication Feature (#1521)

What I did
Radius Management User Authentication Feature

How I did it
HLD: https://github.com/Azure/SONiC/blob/master/doc/aaa/radius_authentication.md

How to verify it
This is the CLI only. The changes are reflected in the Redis Config DB.

Previous command output (if the output of a command-line utility has changed)
New command output (if the output of a command-line utility has changed)
admin@sonic:~$ show radius
RADIUS global auth_type pap (default)
RADIUS global retransmit 3 (default)
RADIUS global timeout 5 (default)
RADIUS global passkey <EMPTY_STRING> (default)

admin@sonic:~$

admin@sonic:~$ sudo config radius
Usage: config radius [OPTIONS] COMMAND [ARGS]...

RADIUS server configuration

Options:
-?, -h, --help Show this message and exit.

Commands:
add Specify a RADIUS server
authtype Specify RADIUS server global auth_type [chap | pap | mschapv2]
default set its default configuration
delete Delete a RADIUS server
nasip Specify RADIUS server global NAS-IP|IPV6-Address
passkey Specify RADIUS server global passkey
retransmit Specify RADIUS server global retry attempts <0 - 10>
sourceip Specify RADIUS server global source ip
statistics Specify RADIUS server global statistics [enable | disable |...
timeout Specify RADIUS server global timeout <1 - 60>
admin@sonic:~$

Blueve committed:
[console] Display success message after line cleared (#1579)

* [console] Display success message after line cleared

Signed-off-by: Jing Kan jika@microsoft.com

ghooo committed:
[config][generic-update] Adding apply-patch, rollback, checkpoints commands (#1536)

#### What I did
Adding apply-patch, rollback, replace, checkpoint, delete-checkpoint, list-checkpoints functionality.

#### How I did it
This PR is implementing the first step in in README.md in the design document: https://github.com/Azure/SONiC/pull/736  

#### How to verify it
Using unit-tests

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

```sh
admin@sonic:~$ sudo config apply-patch --help
Usage: config apply-patch [OPTIONS] PATCH_FILE_PATH

  Apply given patch of updates to Config. A patch is a JsonPatch which
  follows rfc6902. This command can be used do partial updates to the config
  with minimum disruption to running processes. It allows addition as well
  as deletion of configs. The patch file represents a diff of ConfigDb(ABNF)
  format or SonicYang format.

  <patch-file-path>: Path to the patch file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of config of the patch is either
                                  ConfigDb(ABNF) or SonicYang
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.



admin@sonic:~$ sudo config replace --help
Usage: config replace [OPTIONS] TARGET_FILE_PATH

  Replace the whole config with the specified config. The config is replaced
  with minimum disruption e.g. if ACL config is different between current
  and target config only ACL config is updated, and other config/services
  such as DHCP will not be affected. **WARNING** The target config file
  should be the whole config, not just the part intended to be updated.

  <target-file-path>: Path to the target file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of target config is either
                                  ConfigDb(ABNF) or SonicYang
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.



admin@sonic:~$ sudo config rollback --help
Usage: config rollback [OPTIONS] CHECKPOINT_NAME

  Rollback the whole config to the specified checkpoint. The config is
  rolled back with minimum disruption e.g. if ACL config is different
  between current and checkpoint config only ACL config is updated, and
  other config/services such as DHCP will not be affected.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -d, --dry-run   test out the command without affecting config state
  -v, --verbose   print additional details of what the operation is doing
  -?, -h, --help  Show this message and exit.



admin@sonic:~$ sudo config checkpoint --help
Usage: config checkpoint [OPTIONS] CHECKPOINT_NAME

  Take a checkpoint of the whole current config with the specified
  checkpoint name.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -h, -?, --help  Show this message and exit.



admin@sonic:~$ sudo config delete-checkpoint --help
Usage: config delete-checkpoint [OPTIONS] CHECKPOINT_NAME

  Delete a checkpoint with the specified checkpoint name.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -h, -?, --help  Show this message and exit.



admin@sonic:~$ sudo config list-checkpoints --help
Usage: config list-checkpoints [OPTIONS]

  List the config checkpoints available.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -?, -h, --help  Show this message and exit.
```

Commit:
Fix remove ip rif (#1535)

*Added checking of static routes, related to the interface, before deleting of the last IP entry to prevent deleting the RIF if a static route is present in the system.

Signed-off-by: Maksym Belei <Maksym_Belei@jabil.com>

renukamanavalan committed:
disk_check: Check & mount RO as RW using tmpfs (#1569)

What I did
There is a bug that occasionally turn root-overlay as RO. This makes /etc & /home as RO. This blocks any new remote user login, as that needs to write into /etc & /home.

This tool scans /etc & /home (or given dirs) as in RW or RO state. If RO, it could create a writable overlay using tmpfs.
This is transient and stays until next reboot. Any write after the overlay will be lost upon reboot.

But this allows new remote users login.

How I did it
Create upper & work dirs in /run/mount (tmpfs). Mount /etc & /home as lowerdirs and use the same name for final merge. This allows anyone opening a file in /etc or /home to operate on the merged overlay, transparently.

How to verify it
Mount any dir on tmpfs ( mount -t tmpfs tmpfs test_dir)
remount as RO (mount -o remount,ro test_dir)
Pass that dir to this script. (disk_check.py -d ./test_dir)
Now it should be RW

arlakshm committed:
[multi-asic] support show ip bgp neigh/network for multi asic (#1574)

This change is to add support for the commands "show ip bgp neighbor "and "show ip bgp network" for multi asic platforms
Add unit tests for these commands

Signed-off-by: Arvindsrinivasan Lakshmi Narasimhan <arlakshm@microsoft.com>

vadymhlushko-mlnx committed:
Fixed codestyle

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Removed YANG validation from 'config'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Generated from updated YANG model show CLI for PBH & fixed codestyle

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added generated config CLI for PBH

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed codestyle for config

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added 'pbh' subcommand & removed 'hash-field-list' subcommand

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Removed leaf-lists CLI commands, added example of validation click options

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added new data enums from YANG model

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added show pbh ...

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added sceleton for PBH UT

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed show pbh table output

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added 'Symmetric' field to the 'show pbh hash-field' command

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed 'show pbh hash' output

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed 'show pbh rule' output

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed codestyle for show CLI plugin

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added 'show pbh statistics'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added 'flow counter' column to 'show pbh rule'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added sort, changed symmetric check, removed unused check

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added validation for --hash-field-list, --ip-mask

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added is_exist_in_db()

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added is_match, interface_validator

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed codestyle

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
HOT FIX

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed issue with click callbacks

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed sort output for show CLI, typo fixes

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added few 'hash-field' tests

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed issue with 'config pbh rule update'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
HOTFIX for 'config pbh rule update' is_exist_in_db()

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added proper description for --ip-mask

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added ip_mask & hash-field validation for ADD flow

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added tests for 'config pbh hash-field add' flow

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Fixed 'hash-field' add/update, added UT

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added hash UT, used DB migrator

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added UT for 'pbh table', added UT for 'pbh hash' need to add update flow

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added UT for 'pbh hash update'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>

vadymhlushko-mlnx committed:
Added UT for 'pbh rule'

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>
